ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/JumpingGovernor.hs

```diff
@@ -0,0 +1,337 @@
+  | -- | The jumper disagrees with the dynamo and we have found where exactly.
+    -- The last point where the jumper agrees with the dynamo is stored in the
+    -- ClientState.
+    FoundIntersection
+  deriving (Generic)
+
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (JumperState blk)
+
+-- | Instruction from the jumping governor, either to run normal ChainSync, or
+-- to jump to follow the given peer with the given fragment.
+data Instruction blk
+  = RunNormally
+  | JumpTo !(Point blk)
+  deriving (Generic)
+
+deriving instance (HasHeader blk, Eq (Header blk)) => Eq (Instruction blk)
+deriving instance (HasHeader blk, Show (Header blk)) => Show (Instruction blk)
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (Instruction blk)
+
+data JumpResult blk
+  = AcceptedJump !(Point blk)
+  | RejectedJump !(Point blk)
+  deriving (Generic)
+
+deriving instance (HasHeader blk, Eq (Header blk)) => Eq (JumpResult blk)
+deriving instance (HasHeader blk, Show (Header blk)) => Show (JumpResult blk)
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (JumpResult blk)
+
+nextInstruction ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk, HasHeader (Header blk)
+  ) =>
+  Handle m peer blk ->
+  peer ->
+  STM m (Instruction blk)
+nextInstruction governor@Handle {clientStates} peer =
+  ((Map.! peer) <$> readTVar clientStates) >>= \case
+    Dynamo lastJumpSlot -> maybeJump lastJumpSlot >> pure RunNormally
+    Objector _ -> pure RunNormally
+    Jumper nextJumpVar _ _ ->
+      readTVar nextJumpVar >>= \case
+        Nothing -> retry
+        Just fragment -> do
+          writeTVar nextJumpVar Nothing
+          pure $ JumpTo fragment
+  where
+    maybeJump lastJumpSlot = do
+      dynamoFragment <- fromJust <$> getDynamoFragment governor
+      when (succWithOrigin (headSlot dynamoFragment) >= succWithOrigin lastJumpSlot + jumpSize) $ do
+        clientStates' <- readTVar clientStates
+        forM_ (Map.elems clientStates') $ \case
+          Jumper nextJumpVar _ Happy -> writeTVar nextJumpVar $ Just $! castPoint (headPoint dynamoFragment)
+          _ -> pure ()
+
+processJumpResult ::
+  ( MonadSTM m,
+    Ord peer, HasHeader blk, HasHeader (Header blk)
+  ) =>
+  Handle m peer blk ->
+  peer ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult governor@Handle {clientStates, clientHandles} peer jumpResult =
+  ((Map.! peer) <$> readTVar clientStates) >>= \case
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
```

Comment:
What is the meaning of "bumping a candidate fragment" and why should it be bumped?

Reply:
Since we know that the jumper accepts the jump, that means we know their whole candidate fragment is the same as the dynamo's at least up to the jump point. We can therefore set the candidate TVar to that fragment immediately. This will then impact the LoE, GDD, and whatever else reads the ChainSync clients' candidate fragments. That's what I called “bumping” the fragment in my head, but it's just the word that popped in and I'm happy with any other wording!